src/coreclr/gc/gc.cpp

```diff
@@ -22996,13 +22996,21 @@ void gc_heap::mark_phase (int condemned_gen_number, BOOL mark_only_p)
                 int align_const = get_alignment_constant (TRUE);
                 // Pinning the first object in the region.
                 uint8_t* obj_to_pin = heap_segment_mem (gen0_region);
-                pin_by_gc (obj_to_pin);
+                while (obj_to_pin < heap_segment_allocated (gen0_region))
+                {
+                    if (!((CObjectHeader*)obj_to_pin)->IsFree())
+                    {
+                        pin_by_gc (obj_to_pin);
+                        break;
+                    }
+                    obj_to_pin += Align (size (obj_to_pin), align_const);
+                }
 
                 obj_to_pin += Align (size (obj_to_pin), align_const);
```

Comment:
ahh, so I did screw this up when I moved it inside the GC. good thing you are fixing it. now that I'm looking at this code carefully, I see that there's another problem which is obj_to_pin at this line could already be at heap_segment_allocated so we cannot calculate its size. also since now we are kinda duplicating code it might be better to avoid that, ie, something like this: 

```
                int align_const = get_alignment_constant (TRUE);
                // Pinning the first and the middle object in the region.
                uint8_t* boundary = heap_segment_mem (gen0_region);
                uint8_t* obj_to_pin = boundary;
                int num_pinned_objs = 0;
                while (obj_to_pin < heap_segment_allocated (gen0_region))
                {
                    if (obj_to_pin >= boundary && !((CObjectHeader*)obj_to_pin)->IsFree())
                    {
                        pin_by_gc (obj_to_pin);
                        num_pinned_objs++;
                        if (num_pinned_objs >= 2)
                            break;
                        boundary += (gen0_region_size / 2) + 1;
                    }
                    obj_to_pin += Align (size (obj_to_pin), align_const);
                }
```